src/utf8.cpp

```diff
@@ -0,0 +1,346 @@
+    int32_t defaultInFlightBuffer[128], *inFlightBuffer = defaultInFlightBuffer;
+    size_t inFlightBufferLength = 128;
+    std::vector<ScratchpadChunck> scratchPad;
+
+    // A tricky part of this conversion is that the output size is totally unknown, but we want to be efficient with our
+    // allocations. We're going to write the glyph we're normalising in a static buffer (if possible) and write the
+    // filtered, normalized results in a bunch of semi-fixed buffer (the scratchpad) Only when the conversion is over
+    // will we allocate the final buffer and copy everything in there.
+    scratchPad.reserve(8);
+    scratchPad.emplace_back(bufferLength > 1024 ? bufferLength : 1024);
+
+    uint32_t codepoint;
+    uint64_t position = 0;
+    while ((codepoint = fetch_next_codepoint(utf8Buffer, position, bufferLength)) != UTF8_EOF) {
+        // Ignore invalid glyphs
+        if (codepoint == UTF8_INVALID) {
+            continue;
+        }
+
+        size_t decomposedLength = normalize_codepoint(codepoint, inFlightBuffer, inFlightBufferLength);
+
+        if (decomposedLength > inFlightBufferLength) {
+            // We have to increase the size of the in-flight buffer :(
+            if (inFlightBuffer != defaultInFlightBuffer) {
+                free(inFlightBuffer);
+            }
+            inFlightBufferLength = decomposedLength + 64;
+            inFlightBuffer = (int32_t *)malloc((size_t)inFlightBufferLength * sizeof(int32_t));
+            if (inFlightBuffer == nullptr) {
+                return false;
+            }
+
+            // Run the decomposition again with the larger buffer.
+            decomposedLength = normalize_codepoint(codepoint, inFlightBuffer, inFlightBufferLength);
+        }
+
+        // Write the codepoints to the scratchpad
+        for (size_t inflightBufferIndex = 0; inflightBufferIndex < decomposedLength; ++inflightBufferIndex) {
+            char utf8Write[4];
+            uint8_t lengthWritten = write_codepoint((uint32_t)inFlightBuffer[inflightBufferIndex], utf8Write, 4);
+
+            if (scratchPad.back().used + lengthWritten >= scratchPad.back().length) {
+                scratchPad.emplace_back(scratchPad.back().length);
+            }
+
+            ScratchpadChunck &last = scratchPad.back();
+            memcpy(&last.scratchpad[last.used], utf8Write, lengthWritten);
+            last.used += lengthWritten;
+        }
+    }
+
+    // If we had to create a dynamic in-flight buffer, free it on the way out
+    if (inFlightBuffer != defaultInFlightBuffer) {
+        free(inFlightBuffer);
+    }
+
+    // Compile the scratch pads into the final normalized string
+    uint64_t outputLength = 0;
+    for (ScratchpadChunck &chunck : scratchPad) { outputLength += chunck.used; }
+
+    if (outputLength > bufferLength) {
+        *_utf8Buffer = (char *)realloc((void *)*_utf8Buffer, outputLength);
+    }
+
+    uint64_t writeIndex = 0;
+    for (ScratchpadChunck &chunck : scratchPad) {
```

Comment:
If there's only one chunk in the `scratchPad`, we can "move" the contents and avoid the realloc altogether.